Data/Primitive/SmallArray.hs

```diff
@@ -583,20 +593,41 @@ instance Alternative SmallArray where
   some sa | null sa   = emptySmallArray
           | otherwise = die "some" "infinite arrays are not well defined"
 
+data ArrayStack a
```

Comment:
Is it possible to reuse this datatype across both `SmallArray.hs` and `Array.hs`? My (entirely untested) hypothesis is that you could turn this into:

```haskell
data ArrayStack a
  = PushArray !a !(ArrayStack a)
  | EmptyStack
```

and put this somewhere in `Data.Primitive.Internal`.